### PR TITLE
Upgrade to Cliquet 3 (fixes #345, fixes #fixes #369, fixes #421, fixes #424, fixes #433, fixes #434, fixes #461)

### LINKED ISSUE #345: Filter/Sort on subobject properties
Let's say I have the following records:

```
{
    "data": {
        "title": "My title",
        "tags": ["astronomie", "science", "burundi"],
        "author": {
            "username": "natim",
            "first_name": "Rémy"
        }
    }
}
```

**How can I get all records from natim?**

http://localhost:8000/v1/buckets/wiki/collections/articles/records?author.username=natim ?

### LINKED ISSUE #369: Trigger events only when transaction is committed
Currently notifications are triggered when the resource method is executed.

For batch requests, if one of the sub-requests fails, the whole batched is rolledback. However the events are still triggered.

The events should be triggered when the batch transaction is committed.

This will also affect performance since only one event per collection will be triggered for a batch request.

### LINKED ISSUE #421: Add hooks to alter responses/records on events
Currently it is not easy to get a hand on records in a pluggable manner (before reaching storage, or before being rendered in response).

This kind of feature can be useful to introduce some business logic synchronously, like computed fields for example. 

We can start investigating how others solution do it (e.g. [syncano](http://syncano.io/))

### LINKED ISSUE #424: Fix performances obtained with Github auth in tutorial
A single `PUT` on the `default` bucket generates 10 calls to `authentication.unauthenticated_userid()`.

With basic auth it is hard to notice. But when following the Github auth tutorial, it becomes very obvious that something is wrong with the current proposition :)

I started this PR also https://github.com/mozilla-services/cliquet/pull/641

> Note:
> Used this trick to track:
> ```python
>     import traceback
>     for line in traceback.format_stack():
>        print(line.strip())
> ```

### LINKED ISSUE #433: collection_timestamp on an empty collection get updated at each request with postgresql backend.
**Reproduce:**

```
  http GET https://kinto-server.scalingo.io/v1/buckets/default/collections/certificates/records --auth token:my-secret
  Etag: "1455109382839"

  http GET https://kinto-server.scalingo.io/v1/buckets/default/collections/certificates/records --auth token:my-secret
  Etag: "1455109385100"
```

### LINKED ISSUE #434: Default order of records is not -last_modified
While writing kinto-client integration we noticed that creating a list of records using a batch requests didn't preserve the passed order.

### LINKED ISSUE #461: Kinto creates blank record on Content-Type:text/plain
As Kinto is JSON Datastore and whenever we POST text/plain instead of JSON it creates a blank record in DB instead of rejecting text/plain requests. 

Request: 
```
echo '{"data": {"description": "xyz"}}' | http POST http://localhost:8080/v1/buckets/default/collections/test-collection/records Content-Type:text/plain  -v --auth 'token:my-secret'
```

```
POST /v1/buckets/default/collections/test-collection/records HTTP/1.1
Accept: application/json
Accept-Encoding: gzip, deflate
Authorization: Basic dG9rZW46bXktc2VjcmV0
Connection: keep-alive
Content-Length: 33
Content-Type: text/plain
Host: localhost:8080
User-Agent: HTTPie/0.9.2

{"data": {"description": "xyz"}}

``` 

Response :

```
HTTP/1.1 201 Created
Access-Control-Expose-Headers: Retry-After, Content-Length, Alert, Backoff
Connection: keep-alive
Content-Length: 187
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 Feb 2016 10:46:57 GMT
Server: nginx/1.8.0

{
    "data": {
        "id": "1227cf0c-66a5-4620-a178-255e7f479f28",
        "last_modified": 1456417017195
    },
    "permissions": {
        "write": [
            "basicauth:157beded4ea96ad4a21ed7fd91fff2a92a5daf47f054877d2e9768259e61600c"
        ]
    }
}
```

### PR DESCRIPTION
**Pull request by leplatrem**

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -6,11 +6,67 @@ This document describes changes between each past release.
...
 ===================
 
+**Protocol**
+
+- Allow buckets to store arbitrary properties. (#462)
+- Clients are redirected to URLs without trailing slash only if the current URL
+  does not exist (#656)
+- Partial responses can now be specified for nested objects (#445)
+  For example, ``/records?_fields=address.street``.
+- List responses are now sorted by to last_modified descending by default (#434,
```
Comment (almet) [inline CHANGELOG.rst]: "by to"

File: CHANGELOG.rst [hunk 2]
```diff
@@ -6,11 +6,67 @@ This document describes changes between each past release.
...
+- Server now returns 415 error response if client does not send JSON request (#461, mozilla-services/cliquet#667)
+
+**Breaking changes**
+
+- Errors are not swallowed anymore during the execution of ``ResourceChanged``
+  events subscribers.
+
+  Subscribers are still executed within the transaction like before.
+
+  Subscribers are still executed even if transaction is eventually rolledback.
```
Comment (almet) [inline CHANGELOG.rst]: even if **the**

File: CHANGELOG.rst [hunk 3]
```diff
@@ -6,11 +6,67 @@ This document describes changes between each past release.
...
+  Every subscriber execution succeeds, or none.
+
+  Thus, subscribers of these events should only perform operations that are reversed
+  on transaction rollback: most likely database storage operations.
+
+  For irreversible operations see the new ``AfterResourceChanged`` event.
+
+**New features**
+
+- Event subscribers are now ran synchronously and can now alter responses (#421)
```
Comment (almet) [inline CHANGELOG.rst]: s/now/thus

File: CHANGELOG.rst [hunk 4]
```diff
@@ -6,11 +6,67 @@ This document describes changes between each past release.
...
+- Resource events are now merged in batch requests. One event per resource and
+  per action is emitted when a transaction is committed (mozilla-services/cliquet#634)
+- Monitor time of events listeners execution (mozilla-services/cliquet#503)
+- Validate that the client can accept JSON response. (mozilla-services/cliquet#667)
+- Validate that the client can only send JSON request body. (mozilla-services/cliquet#667)
```
Comment (almet) [inline CHANGELOG.rst]: These two are already listed in the protocol changes. Not sure we should repeat them here.

File: CHANGELOG.rst [hunk 5]
```diff
@@ -6,11 +6,67 @@ This document describes changes between each past release.
...
+- Added a new ``AfterResourceChanged`` event, that is sent only when the commit
+  in database is done and successful.
+
+  Subscribers of this event can fail, errors are swallowed and logged. The
+  final transaction result (or response) cannot be altered.
+
+  Since commit occured successfully and operations will not be rolledback,
+  subcribers running irreversible actions should subscribe to this event
+  (like sending messages, deleting files, or run asynchronous tasks).
```
Comment (almet) [inline CHANGELOG.rst]: I'm not sure we should add this level of detail in the CHANGELOG. What about linking to the documentation instead?

**Comment by almet** [comment]
That's a bunch of cool changes! Great.

**Comment by leplatrem** [comment]
I can't reproduce locally the *timestamp* constraint errors of the `loadtest_simulation`.

They seem to always occur on current bucket and collection creation, on the `timestamps_pkey` constraint.
This could be related to this code:
https://github.com/mozilla-services/cliquet/blob/3.0.0/cliquet/storage/postgresql/schema.sql#L154-L158
which inserts a timestamp when could not find one. Unicity can be violated with concurrent transactions.

> PS: good news, load tests have approximately 15% more RPS from cliquet 2.15 to cliquet 3

```
BackendError: IntegrityError: (psycopg2.IntegrityError) duplicate key value violates unique constraint "records_pkey"

DETAIL:  Key (id, parent_id, collection_id)=(blocklist, , bucket) already exists.
```

```
BackendError: IntegrityError: (psycopg2.IntegrityError) duplicate key value violates unique constraint "timestamps_pkey"

DETAIL:  Key (parent_id, collection_id)=(/buckets/blocklist, collection) already exists.

CONTEXT:  SQL statement "INSERT INTO timestamps (parent_id, collection_id, last_modified)

      VALUES (uid, resource, ts)"

PL/pgSQL function collection_timestamp(character varying,character varying) line 14 at SQL statement

 [SQL: '\n        SELECT as_epoch(collection_timestamp(%(parent_id)s, %(collection_id)s))\n            AS last_modified;\n        '] [parameters: {'parent_id': u'/buckets/blocklist', 'collection_id': 'collection'}]; uid=None; errno=None; agent=python-requests/2.9.1; authn_type=None

CRITICAL:venusian:"PUT   /v1/buckets/blocklist/collections/addons" ? (? ms) (psycopg2.IntegrityError) duplicate key value violates unique constraint "timestamps_pkey"

DETAIL:  Key (parent_id, collection_id)=(/buckets/blocklist, collection) already exists.
```

**Comment by leplatrem** [comment]
Would you be willing to merge it like this and go forward ?

**Comment by Natim** [comment]
Yes I would :)

**Comment by almet** [comment]
Yes, I think we should merge this (but it needs to be rebased first)

**Pull request merged**